kratos/includes/linear_master_slave_constraint.h

```diff
@@ -0,0 +1,299 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(LINEAR_MASTER_SLAVE_CONSTRAINT_H)
+#define LINEAR_MASTER_SLAVE_CONSTRAINT_H
+// System includes
+
+// project includes
+#include "includes/define.h"
+#include "includes/master_slave_constraint.h"
+
+namespace Kratos
+{
+  /** @class LinearMasterSlaveConstraint
+    * @ingroup KratosCore
+    * @brief
+    *
+    * This class allows to add a master-slave constraint which is of the form
+    *
+    * SlaveDofVector = T * MasterDofVector + ConstantVector.
+    *
+    * or
+    *
+    * SlaveDof = weight * MasterDof + Constant
+    *
+    * the data T and ConstantVector (or the equivalent scalars) are not stored in the base class, since they can be eventually evaluated runtime.
+    */
+class LinearMasterSlaveConstraint :  public MasterSlaveConstraint
+{
+public:
+    /// Pointer definition of DataValueContainer
+    KRATOS_CLASS_POINTER_DEFINITION(LinearMasterSlaveConstraint);
+    typedef IndexedObject BaseType;
+    typedef std::size_t IndexType;
+    typedef Dof<double> DofType;
+    typedef std::vector< DofType::Pointer > DofPointerVectorType;
+    typedef Node<3> NodeType;
+    typedef std::vector<std::size_t> EquationIdVectorType;
+
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef Kratos::Variable<double> VariableType;
+    typedef Kratos::VariableComponent<Kratos::VectorComponentAdaptor<Kratos::array_1d<double, 3>>> VariableComponentType;
+    ///@name Life Cycle
+    ///@{
+
+    LinearMasterSlaveConstraint(IndexType Id = 0)
+        :
+        MasterSlaveConstraint(Id)
+    {
+    }
+
+    /*
+    * Constructor by passing a vector of Master and slave dofs and corresponding Matrix and constant vector
+    */
+    LinearMasterSlaveConstraint(IndexType Id,
+                                DofPointerVectorType& rMasterDofsVector,
+                                DofPointerVectorType& rSlaveDofsVector,
+                                const MatrixType& rRelationMatrix,
+                                const VectorType& rConstantVector)
+        : MasterSlaveConstraint(Id)
+    {
+        mSlaveDofsVector = rSlaveDofsVector;
```

Comment:
Better to initialize these in the initialization list. By assigning in the body you may prevent move ops from happening automatically.